### ISSUE TITLE
salt-master does not install due to python tornado dependency

### ISSUE THREAD

**Issue by ramonamis** [saltstack/salt#27820]
Upgrading from salt 2015.5.3 (Lithium) to 2015.8.0 (Beryllium) fails on Ubuntu 14.04.3 LTS (GNU/Linux 3.13.0-44-generic x86_64) due to unmet python-tornado dependency. 

First, trying to install salt-master fails:
```
apt-get install salt-master
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 salt-master : Depends: salt-common (= 2015.8.0+ds-1) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

Then, trying to install salt-common fails:
```
apt-get install salt-common=2015.8.0+ds-1
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 salt-common : Depends: python-tornado (>= 4.2.1) but it is not going to be installed
               Recommends: python-mako but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

The version of python-tornado required is not available on apt, and installing via pip does not seem to be a good idea. However, the deb package for tornado can be found on your repo (http://repo.saltstack.com/apt/ubuntu/ubuntu14/pool/main/p/python-tornado/) so as a temporary workaround was added to our repo and installed via apt. 

Since tornado is considered experimental for this release according to the release notes, it should not block installation.

**Comment by eddpayne** [comment]
Just for further information - we have a repo into which we put the salt packages as we need the ability to install different versions which are removed from the public repo when new versions are released. We noticed that with 2015.8 the repo has been split into an ubuntu12 and an ubuntu14 repo, and the naming of the packages now also causes us issues as the python-tornado package has the same name for both distros (rather than having 1trusty1 or similar at the end). Is this change now permanent, in which case we need to also split our repository into ubuntu12 and ubuntu14 repositories?

**Comment by jfindlay** [comment]
@ramonamis, thanks for the report.  Are you sure you are on a 64 bit system, because this is a [known problem](https://github.com/saltstack/salt/issues/27437) on 32 bit systems.  Tornado experimental as a transport mechanism, but it has been factored into other areas of salt, which is why it is a required dependency.

@eddpayne, @dmurphy18 has been constructing the packages and repos and will be better able to answer your questions.  Ultimately the repos should be more standard than they are now.

**Comment by dmurphy18** [comment]
@eddpayne the problem with the python-tornado dependency was fixed before we announced 2015.8.0, and should not be causing any issues now.  Could you have cached the repo before we announced ?.  I shall check to make sure the same fix for Debian 8 was carried forward to Ubuntu in the meantime.

**Comment by ramonamis** [comment]
@jfindlay i am confident that I am using a 64 bit system. 

```
uname -a
Linux deploid2 3.4.4-guest #1 SMP Wed Mar 13 17:43:22 EDT 2013 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by ashutoshnarayan** [comment]
Hi,

Even am using x86_64 bit Oracle Linux System 
Linux <hostname> 2.6.32-431.29.2.el6.x86_64 #1 SMP Tue Sep 9 11:28:47 PDT 2014 x86_64 x86_64 x86_64 GNU/Linux

Am getting same error even after using saltstack repo

Error: Package: python-tornado-4.2.1-1.el6.x86_64 (saltstack-repo)
           Requires: python-backports-ssl_match_hostname
Error: Package: python-urllib3-1.10.2-1.el6.noarch (saltstack-repo)
           Requires: python-backports-ssl_match_hostname

[root@<hostname> yum.repos.d]# pip install --upgrade tornado
/usr/lib/python2.6/site-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
Requirement already up-to-date: tornado in /usr/lib64/python2.6/site-packages
Requirement already up-to-date: backports.ssl-match-hostname in /usr/lib/python2.6/site-packages (from tornado)
Requirement already up-to-date: certifi in /usr/lib/python2.6/site-packages (from tornado)

I have cached the repo yesterday.

Thank you,
Ashutosh Narayan

**Comment by eguven** [comment]
:+1:

**Comment by eddpayne** [comment]
It's possible that it was cached before announcing, once @ramonamis has given me the go-ahead I'll pull the packages and put the back into our repo again (at the moment we've reverted to Lithium in the repo).

Thanks!

**Comment by dmurphy18** [comment]
@ashutoshnarayan @eddpayne Can you test against the latest 2015.8.1 and check if the problem is still occurring for you.

@ashutoshnarayan Please check the repositories enabled in your environment, as python-backports-ssl_match_hostname.noarch is typically available from the extras repository on a RedHat 6 machine, noting you are using Oracle Linux.

**Comment by ashutoshnarayan** [comment]
@dmurphy18 I was able to install latest version ( 2015.8.1 ) salt-master as well as salt-minion on Oracle Linux. I had to check few internal repositories configured for Oracle Linux which was causing dependency problems. Problem is now solved.

Thank you,
Ashutosh Narayan

**Comment by eddpayne** [comment]
I've just checked and I see that the repo is still split into two separate repositories, one called ubuntu14 and one ubuntu12, both of which contain a file named "python-tornado_4.2.1-1_amd64.deb", with different sizes. The issue we had is that as we only have one repository for all of our clients, normally our packages contain strings like 1trusty1 or similar to distinguish between the two distros meaning that we can have both versions in the same repository and the OS will automatically pick the right one. 

In fact, checking further, I see that all the salt packages now have the same name in both ubuntu12 and ubuntu14 repositories so we can only have one or the other version in our repo.

If this split is now permanent we will need to re-factor our internal repository, which could be done if needed, but I'd be interested to know what the reasoning is for the change in naming from the previous version (along the lines of salt-common_2015.5.0+ds-1precise1_all.deb).

**Comment by mathrick** [comment]
@dmurphy18: Tried that, but it fails with `Unknown stable version: 2015.8.1 (valid: 1.6, 1.7, 2014.1, 2014.7, 2015.5, 2015.8, latest)`

**Comment by dmurphy18** [comment]
@mathrick what OS and it's version are you using.  If Redhat 6, package python-backports-ssl_match_hostname is typically provided by the extra's repository.

It will be added in the future to repo.saltstack.com to resolve the issue with Oracle Linux mentioned above.

**Comment by mathrick** [comment]
@dmurphy18 It's the official Ubuntu 14.04 x64 Docker image. It auto-selects 2015.8, and that fails with unmet `python-tornado>=4.0", and trying to force 2015.8.1 results in the error about unknown version.

**Comment by dmurphy18** [comment]
@eddpayne Ubuntu12 and Ubuntu14 are going to remain split, but don't rearrange anything just yet, the layout of repo.saltstack.com is going to be slightly changed in layout soon. However the aim is still to be backwards compatible with the current layout for the immediate future to minimize churn. It would be best to wait till after the layout adjustment

**Comment by dmurphy18** [comment]
@mathrick would it be possible to post the entry in the sources.list for salt, and/or the output of 'apt-cache policy salt-minion', since the python-tornado dependency should be getting satisfied from repo.saltstack.com.

**Comment by mathrick** [comment]
@dmurphy18 I tried rerunning it, and the tornado issue seems to have been resolved. However, a new issue I've been hitting recently is
```
install_salt.sh: 1885: install_salt.sh: wget: not found
gpg: no valid OpenPGP data found.
 * ERROR: Failed to run install_ubuntu_stable_deps()!!!
```
I don't believe it was a problem when I first tried installing salt. It seems that it's been resolved by the bootstrap script getting a release in the meantime (saltstack/salt-bootstrap@35f39898a9e0b13dfffe2913e1db44f10bda08b6). I seem to recall seeing 2015.08.06 as the bootstrapper version back when I wrote my original comment, and it's now bumped to 2015.11.04

**Comment by meggiebot** [comment]
@mathrick it sounds like this issue is resolved, can you open a new issue for your GPG errors?

**Comment by dmurphy18** [comment]
@ramonamis Can you check that with the latest release 2015.8.3 that this issue is resolved and if so, close this issue.

**Comment by antonakv** [comment]
The issue still exists. And why was it closed? A redesign is needed! It's clear. 
That idea to use Python-tornado was quite weird.

# cat /etc/issue
Amazon Linux AMI release 2017.03
Kernel \r on an \m
$ sudo yum clean expire-cache
$ sudo yum update
$ sudo # yum install salt-minion
Loaded plugins: priorities, upgrade-helper
1077 packages excluded due to repository priority protections
Resolving Dependencies
--> Running transaction check
---> Package salt-minion.noarch 0:2017.7.0-1.el6 will be installed
--> Processing Dependency: salt = 2017.7.0-1.el6 for package: salt-minion-2017.7.0-1.el6.noarch
--> Running transaction check
---> Package salt.noarch 0:2017.7.0-1.el6 will be installed
--> Processing Dependency: python27-tornado >= 4.2.1 for package: salt-2017.7.0-1.el6.noarch
--> Processing Dependency: python27-psutil for package: salt-2017.7.0-1.el6.noarch
--> Processing Dependency: PyYAML27 for package: salt-2017.7.0-1.el6.noarch
--> Running transaction check
---> Package PyYAML27.x86_64 0:3.11-2.el6 will be installed
---> Package python27-psutil.x86_64 0:5.2.2-1.ius.el6 will be installed
---> Package salt.noarch 0:2017.7.0-1.el6 will be installed
--> Processing Dependency: python27-tornado >= 4.2.1 for package: salt-2017.7.0-1.el6.noarch
--> Finished Dependency Resolution
Error: Package: salt-2017.7.0-1.el6.noarch (saltstack-repo)
           Requires: python27-tornado >= 4.2.1
           Available: python27-tornado-2.2.1-7.8.amzn1.noarch (amzn-main)
               python27-tornado = 2.2.1-7.8.amzn1

[root@new-prod-balancer2 yum.repos.d]# cat saltstack-repo.repo 
[saltstack-repo]
disabled=False
name=SaltStack repo for RHEL/CentOS 6
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/6/$basearch/latest/SALTSTACK-GPG-KEY.pub
baseurl=https://repo.saltstack.com/yum/redhat/6/$basearch/latest/
humanname=SaltStack repo for RHEL/CentOS 6

Enjoy...

**Comment by dmurphy18** [comment]
@antonakv It appears that you are using the Redhat 6 repository for Amazon Linux.  This is not recommended for versions of Amazon Linux later than 2015.03 when Amazon started utilizing Python 2.7 as the default Python version.  There is a separate tab on the landing page for repo.saltstack.com for Amazon Linux, http://repo.saltstack.com/#amzn which has instructions for installing Salt on Amazon Linux. Salt  packages and dependencies have been built to utilize Amazon Linux and Python 2.7 and using these should resolve the issues you are encountering.  These Salt packages for Amazon Linux were made available starting with the Salt 2016.11.0 release.

Also note that with the release of Salt 2017.7.0, Redhat 6 support has also been updated to utilize Python 2.7 rather the default installed Python 2.6, however is still recommended to utilize the Salt Amazon packages when using Amazon Linux.

Please let me know if this resolves the issues encountered..

**Comment by antonakv** [comment]
Thank you.
It works !
